banyand/liaison/grpc/auth.go

```diff
@@ -0,0 +1,91 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package grpc
+
+import (
+	"context"
+	"strings"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	"github.com/apache/skywalking-banyandb/banyand/liaison/pkg/auth"
+	"github.com/apache/skywalking-banyandb/banyand/liaison/pkg/config"
+)
+
+// authInterceptor gRPC auth interceptor.
+func authInterceptor(ctx context.Context, req any, _ *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+	if !config.Cfg.Enabled {
+		return handler(ctx, req)
+	}
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, status.Errorf(codes.Unauthenticated, "metadata is not provided")
+	}
+	usernameList, usernameOk := md["username"]
+	passwordList, passwordOk := md["password"]
+	if !usernameOk || len(usernameList) == 0 {
+		return nil, status.Errorf(codes.Unauthenticated, "username is not provided correctly")
+	}
+	if !passwordOk || len(passwordList) == 0 {
+		return nil, status.Errorf(codes.Unauthenticated, "password is not provided correctly")
+	}
+	username := usernameList[0]
+	password := passwordList[0]
+
+	for _, user := range config.Cfg.Users {
+		if strings.ReplaceAll(username, " ", "") == strings.ReplaceAll(user.Username, " ", "") &&
+			auth.CheckPassword(strings.ReplaceAll(password, " ", ""), strings.ReplaceAll(user.Password, " ", "")) {
+			return handler(ctx, req)
+		}
+	}
+	return nil, status.Errorf(codes.Unauthenticated, "invalid username or password")
+}
+
+// authStreamInterceptor gRPC auth interceptor for streams.
+func authStreamInterceptor(srv interface{}, ss grpc.ServerStream, _ *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	if !config.Cfg.Enabled {
+		return handler(srv, ss)
+	}
+
+	md, ok := metadata.FromIncomingContext(ss.Context())
+	if !ok {
+		return status.Errorf(codes.Unauthenticated, "metadata is not provided")
+	}
+	usernameList, usernameOk := md["username"]
+	passwordList, passwordOk := md["password"]
+	if !usernameOk || len(usernameList) == 0 {
+		return status.Errorf(codes.Unauthenticated, "username is not provided correctly")
+	}
+	if !passwordOk || len(passwordList) == 0 {
+		return status.Errorf(codes.Unauthenticated, "password is not provided correctly")
+	}
+	username := usernameList[0]
+	password := passwordList[0]
+
+	for _, user := range config.Cfg.Users {
+		if strings.ReplaceAll(username, " ", "") == strings.ReplaceAll(user.Username, " ", "") &&
+			auth.CheckPassword(strings.ReplaceAll(password, " ", ""), strings.ReplaceAll(user.Password, " ", "")) {
+			return handler(srv, ss)
+		}
+	}
+	return status.Errorf(codes.Unauthenticated, "invalid username or password")
```

Comment:
There are duplicated pieces of code here. Please create a function to reuse the same logic.